src/plugins/event-plugins/ResourcePlugin.ts

```diff
@@ -48,83 +48,75 @@ export class ResourcePlugin extends InternalPlugin {
+                // PutRumEvents.
+                isPutRumEventsCall(name, this.context.config.endpointUrl.host)
+            ) {
+                continue;
+            }
 
-        // Record events from resources in sample until we hit the resource limit
-        shuffle(sample);
-        while (sample.length > 0 && this.eventCount < this.config.eventLimit) {
-            this.recordResourceEvent(sample.pop() as PerformanceResourceTiming);
-            this.eventCount++;
+            // Sampling logic
+            const fileType = getResourceFileType(initiatorType);
+            if (this.config.recordAllTypes.includes(fileType)) {
+                // Always record
+                this.recordResourceEvent(entry);
+            } else if (
+                this.sampleCount < this.config.eventLimit &&
+                this.config.sampleTypes.includes(fileType)
+            ) {
+                // Only sample first N
+                this.recordResourceEvent(entry);
+                this.sampleCount++;
+            }
         }
     };
 
-    recordResourceEvent = ({
-        name,
-        startTime,
-        initiatorType,
-        duration,
-        transferSize
-    }: PerformanceResourceTiming): void => {
-        const pathRegex =
-            /.*\/application\/[0-9a-f]{8}-[0-9a-f]{4}-[0-9a-f]{4}-[0-9a-f]{4}-[0-9a-f]{12}\/events/;
-        const entryUrl = new URL(name);
-        if (
-            entryUrl.host === this.context.config.endpointUrl.host &&
-            pathRegex.test(entryUrl.pathname)
-        ) {
-            // Ignore calls to PutRumEvents (i.e., the CloudWatch RUM data
-            // plane), otherwise we end up in an infinite loop of recording
-            // PutRumEvents.
-            return;
-        }
-
-        if (this.context?.record) {
-            const eventData: ResourceEvent = {
-                version: '1.0.0',
-                initiatorType,
-                startTime,
-                duration,
-                fileType: getResourceFileType(name, initiatorType),
-                transferSize
-            };
-            if (this.context.config.recordResourceUrl) {
-                eventData.targetUrl = name;
-            }
-            this.context.record(PERFORMANCE_RESOURCE_EVENT_TYPE, eventData);
-        }
+    recordResourceEvent = (r: PerformanceResourceTimingPolyfill): void => {
+        this.context?.record(PERFORMANCE_RESOURCE_EVENT_TYPE, {
+            name: this.context.config.recordResourceUrl ? r.name : undefined,
+            entryType: RESOURCE,
+            startTime: r.startTime,
+            duration: r.duration,
+            connectStart: r.connectStart,
+            connectEnd: r.connectEnd,
+            decodedBodySize: r.decodedBodySize,
+            domainLookupEnd: r.domainLookupEnd,
+            domainLookupStart: r.domainLookupStart,
+            fetchStart: r.fetchStart,
+            encodedBodySize: r.encodedBodySize,
+            initiatorType: r.initiatorType,
```

Comment:
question: with the schema update, the initiatorType has to be one of the [21 valid initiatorTypes](https://developer.mozilla.org/en-US/docs/Web/API/PerformanceResourceTiming/initiatorType). in the case that there are new initiatorTypes that are added, events with the new initiatorTypes will not be ingested. could be out of scope for this PR but we might need to handle this edge case by adding a catch-all value (e.g. "other") that is considered to be a valid initiatorType so that events aren't dropped until we update the schema with the new initiatorType.

Reply:
+1 https://github.com/aws-observability/aws-rum-web/pull/489#issuecomment-1897503723